In [1]:
#importing libraries
import numpy as np
import pandas as pd


### Question 1: Some of the orders are stored in another csv file named `bigkart_newsales`. Read the csv file, store it in a data frame and add it to the `bigkart_sales` data frame. Find the total sales value of the category 'Office Supplies' after combining the dataframes 

In [30]:
bigkart_sales = pd.read_csv('bigkart_sales.csv')
bigkart_newsales = pd.read_csv('bigkart_newsales.csv')
# bigkart_sales = bigkart_sales.merge(bigkart_newsales, on = 'Order ID', how = 'outer')

In [31]:
bigkart_sales.shape

(59, 8)

In [32]:
dataframes = [bigkart_sales, bigkart_newsales]

main_frame = pd.concat(dataframes,axis=0,ignore_index=True, sort=False)
main_frame.groupby(by='Category').Sales.sum()

Category
Furniture          2460
Office Supplies    7970
Technology         8588
Name: Sales, dtype: int64

#### So the total sales value of the category 'Office Supplies' is 7970


### Question 2: There are some duplicate rows in the data frame. Drop these rows and calculate the total sales value of the category Office Supplies.

In [33]:
main_frame.drop_duplicates(subset=None,keep='first',inplace = True)
main_frame.groupby(by='Category').Sales.sum()

Category
Furniture          2460
Office Supplies    6964
Technology         7494
Name: Sales, dtype: int64

#### After dropping the duplicate rows. the total sales value of the category Office Supplies is 6964.

### Question 3: Find the most profitable category and sub category combination based on the net profit.

In [34]:
main_frame.groupby(by=['Category','Sub-Category']).Profit.sum()

Category         Sub-Category
Furniture        Bookcases        308
                 Chairs           -49
                 Furnishings       74
Office Supplies  Appliances       229
                 Art              275
                 Binders          156
                 Envelopes         35
                 Fasteners         10
                 Labels            30
                 Paper             15
                 Storage          192
                 Supplies         103
Technology       Accessories      215
                 Copiers            0
                 Machines         168
                 Phones          1618
Name: Profit, dtype: int64

#### The most profitable category and sub category combination based on net profit is Technology-Phones

### Question 4: How many invalid order IDs are there in the data frame. An order id is of the form AZ-2011-Y where Y represents a whole number. A Order ID is said to be valid only if Y consists of 7 digits. Find the number of invalid order order IDs in the data frame.

In [35]:
main_frame['Order no']=main_frame['Order ID'].str.split('-')
main_frame['Order no']=main_frame['Order no'].apply(lambda x: x[2])
main_frame['no_of_digits']=main_frame['Order no'].apply(lambda x: len(x))
main_frame['no_of_digits'].value_counts()

7    54
6     7
Name: no_of_digits, dtype: int64

#### The number of invalid order id is 7.

### Question 5: Find the top 25 orders based on sales value and find the number of orders which belong to furniture category.

In [36]:
top_25=main_frame.sort_values(by='Sales', ascending=False)
top_25=top_25.iloc[:25,]
top_25['Category'].value_counts()

Office Supplies    11
Technology          9
Furniture           5
Name: Category, dtype: int64

#### So in the top 25 orders based on sales value, Office Supplies category has 11 orders, Technology category has 9 orders and Furniture category has 5 orders.

### Question 6: Among the orders with sales>250 and profit>50, find the product name of the fourth highest order based on sales value.

In [40]:
highest_order=main_frame.loc[(main_frame['Sales']>250) & (main_frame['Profit']>50)]
highest_order=highest_order.sort_values(by='Sales',ascending=False) 
print(highest_order)

           Order ID                         Product Name  Discount  Sales  \
30  AZ-2011-1410648         Nokia Smart Phone, Full Size       0.0   1908   
58   AZ-2011-176674                Hoover Microwave, Red       0.1   1667   
8   AZ-2011-1174243     Nokia Audio Dock, with Caller ID       0.0   1334   
20  AZ-2011-1322840     Motorola Headset, with Caller ID       0.0    957   
3   AZ-2011-1372644           Panasonic Printer, Durable       0.0    800   
18  AZ-2011-1406494         Fellowes Lockers, Industrial       0.1    748   
39  AZ-2011-1536006      Logitech Keyboard, Programmable       0.0    666   
33  AZ-2011-1445262          Apple Smart Phone, Cordless       0.0    636   
31   AZ-2011-144325        Bush Stackable Bookrack, Pine       0.0    630   
14  AZ-2011-1260928        Eldon File Cart, Single Width       0.1    576   
12  AZ-2011-1253407       Safco Stackable Bookrack, Pine       0.1    541   
48  AZ-2011-1672552      Binney & Smith Sketch Pad, Blue       0.0    510   

#### The product name of the fourth highest order based on sales value is 'Motorola Headset, with Caller ID'

### Question 7: Remove the orders with negative profit by dropping the corresponding rows with negative `Profit`. Find the product that makes the lowest profit per Quantity in the Technology category.

In [38]:
lowest_profit=main_frame[main_frame.Profit>0]
lowest_profit=lowest_profit.loc[lowest_profit.Category=='Technology']
lowest_profit['Profit_per_qty']=lowest_profit['Profit']/lowest_profit['Quantity']
lowest_profit.sort_values(by='Profit_per_qty',ascending=True)

Order ID                      Product Name  Discount  Sales  \
39  AZ-2011-1536006   Logitech Keyboard, Programmable       0.0    666   
2   AZ-2011-1087704     Belkin Flash Drive, Bluetooth       0.0    294   
63  AZ-2011-1967754      Logitech Numeric Keypad, USB       0.0     93   
8   AZ-2011-1174243  Nokia Audio Dock, with Caller ID       0.0   1334   
20  AZ-2011-1322840  Motorola Headset, with Caller ID       0.0    957   
3   AZ-2011-1372644        Panasonic Printer, Durable       0.0    800   
52  AZ-2011-1722024            Cisco Audio Dock, VoIP       0.0    364   
33  AZ-2011-1445262       Apple Smart Phone, Cordless       0.0    636   
30  AZ-2011-1410648      Nokia Smart Phone, Full Size       0.0   1908   

    Profit  Quantity    Category Sub-Category Order no  no_of_digits  \
39      66         9  Technology  Accessories  1536006             7   
2      109         7  Technology  Accessories  1087704             7   
63      40         2  Technology  Accessories  1967754             7   
8      200         8  Technology       Phones  1174243             7   
20     316        12  Technology       Phones  1322840             7   
3      168         3  Technology     Machines  1372644             7   
52     142         2  Technology       Phones  1722024             7   
33     140         1  Technology       Phones  1445262             7   
30     820         3  Technology       Phones  1410648             7   

    Profit_per_qty  
39        7.333333  
2        15.571429  
63       20.000000  
8        25.000000  
20       26.333333  
3        56.000000  
52       71.000000  
33      140.000000  
30      273.333333

#### After removing product with negative profit, The product that makes the lowest profit per quantity in the Technology category is 'Logitech Keyboard, Programmable' with profit per quantity value of '7.333333'